### DISCUSSION TITLE
Bindgen on MacOS 14 fails to link dylib with lib not found error

### DISCUSSION THREAD

**Discussion by marvin-hansen** [rust-lang/rust-bindgen#2673] [Q&A]
Macos: 14 / M1
rustc 1.73.0 (cc66ad468 2023-10-03)
Homebrew clang version 17.0.3
Target: arm64-apple-darwin23.0.0
Thread model: posix
InstalledDir: /opt/homebrew/opt/llvm/bin

Hi, 

I'm working on a Rust wrapper for quaser DB that has a C API for all FFI clients. 
The entire C api is wrapped in one libqdb_api.dylib

I am using Bindgen to generate the Rust bindings and that works well meaning I got all bindings generated and the project compiles. 

However, the tests don't runs because the C headerfiles rely on linking against libqdb_api and somehow I cannot convince cargo to find this lib required for linking. Also, running a simple example code fails with the same error.

         = note: ld: warning: ignoring duplicate libraries: '-lc', '-lm'

          ld: library 'libqdb_api' not found

          clang: error: linker command failed with exit code 1 (use -v to see invocation)

I know that I cannot set the DYLD_LIBRARY_PATH as an env variable because of OSX [System Integrity Protection](https://developer.apple.com/library/archive/documentation/Security/Conceptual/System_Integrity_Protection_Guide/RuntimeProtections/RuntimeProtections.html#//apple_ref/doc/uid/TP40016462-CH3-SW1) (SIP). Rather, [in a similar issue,](https://github.com/rust-lang/cargo/issues/4421) it was suggested to use the instal_name_tool instead and so I tried the following:

      install_name_tool -change @rpath/libqdb_api.dylib "/Users/marvin/CLionProjects/quasar-rs/qdb/lib/libqdb_api.dylib" target/debug/libqdb_api.dylib

Which is supposed to switch the rpath to the target folder since the build script is executed relative to the target folder. However, this does not worked either for me. 

The question is: How do I configure cargo to find the dylib, which is in the project folder?

Technically, I can copy the  libqdb_api.dylib into any other folder, and I actually tried it but nothing worked so I believe my configuration must be wrong. 

Any help is appreciated. 

What I did:

* Copied the C API into the root folder of the project
   -relative path is: qdb/lib
* Added bindgen as a dev dependency to cargo.toml.
* Installed llvm via brew
* Set LLVM_CONFIG_PATH & LIBCLANG_PATH in build.rs
* Wrote a wrapper file to include headers for generate bindings
* Wrote a build file, see below:

Log files:
* [Build log](https://github.com/DexStream-ai/quasar-rs/blob/main/build.txt)
* [Error log ](https://github.com/DexStream-ai/quasar-rs/blob/main/error.txt)

[Github repo](https://github.com/DexStream-ai/quasar-rs)

```rust
use std::env;
use std::path::{Path, PathBuf};

fn main() {
    // Tried this to no avail. https://github.com/rust-lang/cargo/issues/4421
    // install_name_tool -change @rpath/libqdb_api.dylib "/Users/marvin/CLionProjects/quasar-rs/qdb/lib/libqdb_api.dylib" target/debug/libqdb_api.dylib

    // Set LLVM config & libclang path https://rust-lang.github.io/rust-bindgen/requirements.html#clang
    let key = "LLVM_CONFIG_PATH";
    env::set_var(key, env::var(key).unwrap_or("/opt/homebrew/opt/llvm/bin/llvm-config".to_string()));

    let key = "LIBCLANG_PATH";
    env::set_var(key, env::var(key).unwrap_or("/opt/homebrew/opt/llvm/lib".to_string()));

    // linking c to the underlying dylib library https://www.reddit.com/r/rust/comments/885t1h/bindgen_linking_question/
    println!("cargo:rustc-link-lib=dylib=c++");
    println!("cargo:rustc-link-lib=dylib=c++abi");
    println!("cargo:rustc-link-lib=dylib=libqdb_api");

    // Set the dylib search path relative to the current crate
    // https://stackoverflow.com/questions/41917096/how-do-i-make-rustc-link-search-relative-to-the-project-location
    let dir = env::var("CARGO_MANIFEST_DIR").unwrap();
    println!("cargo:rustc-link-search=native={}", Path::new(&dir).join("qdb/lib").display());

    // Tell cargo to invalidate the built crate whenever the wrapper changes
    println!("cargo:rerun-if-changed=wrapper.h");

    // The bindgen::Builder is the main entry point to bindgen,
    // and lets you build up options for the resulting bindings.
    let bindings = bindgen::Builder::default()
        // The input header we would like to generate bindings for.
        .header("wrapper.h")
        // Derive debug implementation for structs and enums
        .derive_debug(true)
        // Tell cargo to invalidate the built crate whenever any of the header files changed.
        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
        // Finish the builder and generate the bindings.
        .generate()
        // Unwrap the Result and panic on failure.
        .expect("Unable to generate bindings");

    // Write the bindings to the $OUT_DIR/bindings.rs file.
    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

**Comment by edouarda** [comment]
Hi, it seems you're linking an ARM64 executable against an Intel x64 library. We're working on making the OSX ARM64 library available.

**Comment by pvdrz** [comment]
Closed as it seems this is not a bindgen issue
